recirq/quantum_chess/circuit_transformer.py

```diff
@@ -181,27 +193,32 @@ def map_helper(self,
         # This makes back-tracking easier.
         node_to_map = nodes_to_map[0]
 
+
         for node_to_try in valid_adjacent_qubits:
             # Add proposed qubit to the mapping
             # and remove it from available qubits
             mapping[node_to_map] = node_to_try
             available_qubits.remove(node_to_try)
-
+            nodes_trying.append(node_to_map)
```

Comment:
Isn't this just the inverse of available_qubits?  Why do we need the extra variable?

Reply:
In a typical DFS we don't need this, but what's special here is:

# call this the 1st run
success = self.map_helper(node_to_map, mapping, available_qubits,
                                      graph, nodes_trying)
# call this the 2nd run
if success:
                success = self.map_helper(cur_node, mapping, available_qubits,
                                          graph, nodes_trying)

When the first run returns success, but the second run returns failure, according to the logic before this PR only the top-level nodes (i.e. node_to_try and node_to_map) will be released from mapping and added back into the available_qubits, but actually the nodes involved in the first call also need to released/added back.

The last test in circuit_transformer_test.py is a good example of this case.

Reply:
In a normal DFS we don't need this extra variable. But what's special here is that 

# call this the 1st run
            success = self.map_helper(node_to_map, mapping, available_qubits,
                                      graph, nodes_trying)
# call this the 2nd run
            if success:
                success = self.map_helper(cur_node, mapping, available_qubits,
                                          graph, nodes_trying)

When the 1st run returns success but the second run returns failure, all nodes involved in 1st and 2nd run are supposed to be reversed in mapping/available_qubits, while before this PR the behavior is to only reverse the top-level ones (i.e. node_to_map, node_to_try). The result is that the loop after this will have less available_qubits to try for, and may end up with a failure in creating the mapping. One perfect example is the last testing case in circuit_transformer_test.py.

Reply:
Would it be possible to document this or explain it somewhere?  (Either in a comment/docstring or in the PR description)?

Reply:
Thanks Doug. It's both added in the PR description and the docstring.